File: algorithms/graph/02_BFS.java
import java.util.*;

/**
 * BFS — Breadth First Search
 *
 * IDEA: Visit all neighbors at distance 1 first, then distance 2, then 3...
 *       Like ripples in water — expands level by level.
 *
 * TOOL: Queue (FIFO) — process nodes in the order you discover them
 *
 * USE CASES:
 *   ✅ Shortest path in UNWEIGHTED graph (fewest edges)
 *   ✅ Level order traversal
 *   ✅ Find all nodes reachable from a source
 *   ✅ Check if graph is connected
 *   ✅ "minimum steps" problems
 *
 * TIME:  O(V + E) — visit every vertex and edge once
 * SPACE: O(V)     — queue + visited set
 *
 * Run: javac 02_BFS.java && java BFS
 */
class BFS {

    // ─── Build graph helper ───────────────────────────────────────────────────
    static Map<Integer, List<Integer>> buildGraph() {
        // Graph:   1—2—5
        //          |   |
        //          3—4—6
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[][] edges = {{1,2},{1,3},{2,5},{3,4},{4,6},{5,6}};
        for (int[] e : edges) {
            graph.computeIfAbsent(e[0], k -> new ArrayList<>()).add(e[1]);
            graph.computeIfAbsent(e[1], k -> new ArrayList<>()).add(e[0]);
        }
        return graph;
    }

    // ─── 1. Basic BFS traversal ───────────────────────────────────────────────
    // Visits every reachable node, prints in BFS order
    static void bfsTraversal(Map<Integer, List<Integer>> graph, int start) {
        System.out.println("=== BFS TRAVERSAL from node " + start + " ===");

        Queue<Integer> queue   = new LinkedList<>();
        Set<Integer>   visited = new HashSet<>();

        queue.add(start);
        visited.add(start);

        while (!queue.isEmpty()) {
            int node = queue.poll();           // take from FRONT
            System.out.print(node + " ");

            for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);     // mark BEFORE adding to queue
                    queue.add(neighbor);       // add to BACK
                }
            }
        }
        System.out.println();
    }

    // ─── 2. BFS level by level ───────────────────────────────────────────────
    // Shows which nodes are at distance 1, 2, 3... from start
    static void bfsLevels(Map<Integer, List<Integer>> graph, int start) {
        System.out.println("\n=== BFS LEVELS from node " + start + " ===");

        Queue<Integer> queue   = new LinkedList<>();
        Set<Integer>   visited = new HashSet<>();

        queue.add(start);
        visited.add(start);
        int level = 0;

        while (!queue.isEmpty()) {
            int size = queue.size(); // number of nodes at THIS level
            System.out.print("Level " + level + ": ");

            for (int i = 0; i < size; i++) {   // process exactly this level
                int node = queue.poll();
                System.out.print(node + " ");

                for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                    if (!visited.contains(neighbor)) {
                        visited.add(neighbor);
                        queue.add(neighbor);
                    }
                }
            }
            System.out.println();
            level++;
        }
    }

    // ─── 3. BFS shortest path ────────────────────────────────────────────────
    // Returns the shortest path (fewest edges) from start to end
    // Works only for UNWEIGHTED graphs
    static List<Integer> shortestPath(Map<Integer, List<Integer>> graph, int start, int end) {
        System.out.println("\n=== SHORTEST PATH from " + start + " to " + end + " ===");

        Queue<Integer>        queue   = new LinkedList<>();
        Map<Integer, Integer> parent  = new HashMap<>(); // child → parent (to reconstruct path)

        queue.add(start);
        parent.put(start, -1); // start has no parent

        while (!queue.isEmpty()) {
            int node = queue.poll();

            if (node == end) break; // found it — stop early

            for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (!parent.containsKey(neighbor)) { // not visited
                    parent.put(neighbor, node);
                    queue.add(neighbor);
                }
            }
        }

        // Reconstruct path by walking back from end → start via parent map
        if (!parent.containsKey(end)) {
            System.out.println("No path found");
            return new ArrayList<>();
        }

        List<Integer> path = new ArrayList<>();
        for (int cur = end; cur != -1; cur = parent.get(cur)) {
            path.add(cur);
        }
        Collections.reverse(path);
        System.out.println("Path: " + path + "  (length: " + (path.size() - 1) + " edges)");
        return path;
    }

    // ─── 4. BFS shortest distance to all nodes ───────────────────────────────
    static void shortestDistances(Map<Integer, List<Integer>> graph, int start) {
        System.out.println("\n=== SHORTEST DISTANCES from node " + start + " ===");

        Queue<Integer>        queue    = new LinkedList<>();
        Map<Integer, Integer> distance = new HashMap<>();

        queue.add(start);
        distance.put(start, 0);

        while (!queue.isEmpty()) {
            int node = queue.poll();

            for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (!distance.containsKey(neighbor)) {
                    distance.put(neighbor, distance.get(node) + 1);
                    queue.add(neighbor);
                }
            }
        }

        for (int node : distance.keySet()) {
            System.out.println("  Node " + node + " → distance " + distance.get(node));
        }
    }

    // ─── 5. BFS step-by-step trace ───────────────────────────────────────────
    // See exactly what happens inside the queue — good for learning
    static void bfsTrace(Map<Integer, List<Integer>> graph, int start) {
        System.out.println("\n=== BFS INTERNAL TRACE from node " + start + " ===");
        System.out.println("(watching the queue at each step)\n");

        Queue<Integer> queue   = new LinkedList<>();
        Set<Integer>   visited = new HashSet<>();

        queue.add(start);
        visited.add(start);

        while (!queue.isEmpty()) {
            System.out.println("  Queue: " + queue);
            int node = queue.poll();
            System.out.println("  Dequeue → process node " + node);

            for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.add(neighbor);
                    System.out.println("    Enqueue neighbor " + neighbor);
                }
            }
        }
    }

    public static void main(String[] args) {
     //   Map<Integer, List<Integer>> graph = buildGraph();

        Map<Integer, List<Integer>> graph = new HashMap<>();
        
        int[][] edges = {{1,2},{1,3},{2,5},{3,4},{4,6},{5,6}};
       
        for (int[] e : edges) {
            graph.computeIfAbsent(e[0], k -> new ArrayList<>()).add(e[1]);
            graph.computeIfAbsent(e[1], k -> new ArrayList<>()).add(e[0]);
        }

        // bfsTraversal(graph, 1);
        // bfsLevels(graph, 1);
        // shortestPath(graph, 1, 6);
        // shortestPath(graph, 1, 4);
        // shortestDistances(graph, 1);
        // bfsTrace(graph, 1);

        // System.out.println("\n=== KEY POINTS TO REMEMBER ===");
        // System.out.println("""
        //     1. Always mark visited BEFORE adding to queue (not after polling)
        //        → If you mark after polling, same node gets added to queue multiple times

        //     2. Use Queue (LinkedList) — NOT Stack
        //        → Queue = FIFO = level by level = BFS
        //        → Stack = LIFO = deep first = DFS

        //     3. BFS gives shortest path only for UNWEIGHTED graphs
        //        → For weighted graphs → Dijkstra

        //     4. Level trick: int size = queue.size() at start of each level iteration
        //        → This freezes the count of nodes at current level
        //     """);
    }
}

File: algorithms/graph/python/02_bfs.py
# ============================================================
# CONCEPT 2: BFS — Breadth First Search
#
# BFS = explore level by level (like ripples in water)
# Best for: MINIMUM NUMBER OF STEPS/HOPS/JUMPS
#
# Uses a QUEUE (first in, first out)
# Why queue? Because you want to process level 1 before level 2
#
# Template:
#   queue = [start]
#   visited = {start}
#   steps = 0
#   while queue:
#       for each node in current level:
#           if node == target: return steps
#           add unvisited neighbors to queue
#       steps += 1
# ============================================================

from collections import deque


# ============================================================
# QUESTION 1: Can you reach the destination?
#
# Graph (directed):
#   0 → 1, 2
#   1 → 3
#   2 → 4
#   3 → 4
#   4 → (nothing)
#
# Using BFS, find if you can reach node 4 from node 0.
# ============================================================
print("QUESTION 1: Can you reach node 4 from node 0?")
print("="*40)

graph = {
    0: [1, 2],
    1: [3],
    2: [4],
    3: [4],
    4: [],
}

def can_reach(graph, start, target):
    queue   = deque([start])
    visited = {start}

    while queue:
        node = queue.popleft()
        print(f"  Visiting: {node}")

        if node == target:
            return True

        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)

    return False

result = can_reach(graph, 0, 4)
print(f"\nCan reach 4? {result}")

# NUANCE: visited set prevents infinite loops.
# Without it, you'd revisit nodes forever in cycles.


input("\nPress Enter for Question 2...\n")


# ============================================================
# QUESTION 2: Minimum hops to reach destination
#
# Same graph. How many hops (edges) from 0 to 4?
#
# Answer should be 2: 0 → 2 → 4
#
# KEY: BFS explores level by level, so the FIRST TIME
#      you reach a node, that's the shortest path.
# ============================================================
print("QUESTION 2: Minimum hops from 0 to 4")
print("="*40)

def min_hops(graph, start, target):
    queue   = deque([(start, 0)])  # (node, hops_so_far)
    visited = {start}

    while queue:
        node, hops = queue.popleft()
        print(f"  Visiting: {node}  hops so far: {hops}")

        if node == target:
            return hops

        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append((neighbor, hops + 1))

    return -1  # not reachable

result = min_hops(graph, 0, 4)
print(f"\nMin hops: {result}")

# NUANCE: We track hops_so_far in the queue alongside the node.
# When we first reach target, that's the minimum — BFS guarantees this.


input("\nPress Enter for Question 3...\n")


# ============================================================
# QUESTION 3: Print the actual path (not just hop count)
#
# Instead of just counting hops, track WHERE WE CAME FROM.
# Use a "parent" dict: parent[node] = node_we_came_from
# Then trace back from target to start.
# ============================================================
print("QUESTION 3: Print the actual shortest path")
print("="*40)

def shortest_path(graph, start, target):
    queue   = deque([start])
    visited = {start}
    parent  = {start: None}   # ← tracks how we got here

    while queue:
        node = queue.popleft()

        if node == target:
            # trace back
            path = []
            while node is not None:
                path.append(node)
                node = parent[node]
            return list(reversed(path))

        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                parent[neighbor] = node   # ← came from `node`
                queue.append(neighbor)

    return []

path = shortest_path(graph, 0, 4)
print(f"Shortest path: {' → '.join(map(str, path))}")

# NUANCE: parent dict is the key trick for path reconstruction.
# Very common in interviews — always know this.


input("\nPress Enter for Question 4...\n")


# ============================================================
# QUESTION 4: Grid BFS — most common interview form
#
# Given a 4x4 grid. Start at (0,0), reach (3,3).
# '#' = wall (blocked), '.' = open
# Find minimum steps.
#
# grid:
#   . . . .
#   . # # .
#   . # . .
#   . . . .
# ============================================================
print("QUESTION 4: Grid BFS — minimum steps")
print("="*40)

grid = [
    ['.', '.', '.', '.'],
    ['.', '#', '#', '.'],
    ['.', '#', '.', '.'],
    ['.', '.', '.', '.'],
]

def grid_bfs(grid, start, target):
    rows, cols = len(grid), len(grid[0])
    queue   = deque([(start, 0)])
    visited = {start}
    directions = [(0,1),(0,-1),(1,0),(-1,0)]  # right, left, down, up

    while queue:
        (r, c), steps = queue.popleft()

        if (r, c) == target:
            return steps

        for dr, dc in directions:
            nr, nc = r+dr, c+dc
            if 0 <= nr < rows and 0 <= nc < cols:
                if grid[nr][nc] != '#' and (nr, nc) not in visited:
                    visited.add((nr, nc))
                    queue.append(((nr, nc), steps + 1))

    return -1

result = grid_bfs(grid, (0,0), (3,3))
print(f"Min steps (0,0) → (3,3): {result}")

print("\n✅ KEY TAKEAWAYS:")
print("  - BFS = min hops (not min cost — that's Dijkstra)")
print("  - Always use a visited set to avoid revisiting")
print("  - Store (node, steps) or (node, path) in queue")
print("  - parent dict = reconstruct path")
print("  - Grid BFS = same idea, neighbors are up/down/left/right")
